### PR TITLE
Add LockSingleton for ECC lock

### PR DESCRIPTION
**Pull request by posterzh**
**Why**
To prevent ECC test fail, add the LockSingleton.  Use this singleton lock whenever the ECC is accessed.

**How**
Using Posix IPC, create a system wide semaphore to lock the ECC.
Also let hm-pyhelper provide convenient `@ecc_lock` decorator to keep the main logic simple and neat.

**References**
https://github.com/NebraLtd/hm-diag/issues/167

### CODE REVIEW

**Comment by marvinmarnold** [comment]
@posterzh the demos are nice but we should have unit tests for all new code, especially things this complex.

I took the liberty of [adding two](cc64fff) (sorry for breaking flake8). `test_ecc_lock` is working correctly. But `test_lock_singleton_failure` is hanging for an unknown reason when running `pytest hm_pyhelper/tests/test_lock_singleton.py`. Can you help me debug? Can you then finish implementing the test (you should just be able to uncomment the rest of the code).?

**Comment by posterzh** [comment]
@marvinmarnold The reason for the hangout is because there are no resources available. I updated the `LockSingleton` constructor to set the count of resources explicitly if it's set in the parameter.
And I added 2 test cases for `LockSingleton`, `test_lock_singleton_basic` and `test_lock_singleton_racing`. The result of the `test_lock_singleton_racing` looks like below.
image

File: hm_pyhelper/lock_singleton.py [hunk 1]
```diff
@@ -11,17 +11,30 @@
...
         old_umask = os.umask(0)
         try:
             self._sem = posix_ipc.Semaphore(self._name,
                                             mode=self._mode,
                                             flags=posix_ipc.O_CREAT,
-                                            initial_value=initial_value)
+                                            initial_value=self._resource_count)
+            # Override the resource count if it's explicitly set. Otherwise the previous value
+            # of the resource count is preserved.
+            if resource_count and resource_count != self._sem.value:
```
Comment (marvinmarnold) [inline hm_pyhelper/lock_singleton.py:32]: This code smells funny to me. Did you find any documentation online that say this is a common pattern?

File: hm_pyhelper/lock_singleton.py [hunk 2]
```diff
@@ -11,17 +11,30 @@
 class LockSingleton(object):
```
Comment (marvinmarnold) [inline hm_pyhelper/lock_singleton.py:11]: It doesn't seem like `LockSingleton` is still a good name for this. It is now possible to create many different instances. I think `IpcLock` or `InterprocessLock` are more reflective of it now.

File: hm_pyhelper/lock_singleton.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    wrong IPC object or whatever internal issue.
+    """
+    def __init__(self, *args, **kwargs):
+        pass
+
+
+def ecc_lock(func):
+    """Returns an ECC LOCK decorator.
+    """
+    lock = LockSingleton(LOCK_ECC)
```
Comment (marvinmarnold) [inline hm_pyhelper/lock_singleton.py:87]: This should have a timeout value of 1-2seconds.

File: hm_pyhelper/lock_singleton.py (hunk 1 above)
Comment (marvinmarnold) [reply hm_pyhelper/lock_singleton.py:32]: What value do we even get in wrapping `posix_ipc.Semaphore`? Why doesn't ecc_lock just use `posix_ipc.Semaphore` directly?

File: hm_pyhelper/tests/test_lock_singleton.py [hunk 4]
```diff
@@ -69,32 +72,34 @@ def fast_task():
         fast_thread.start()
 
     def test_lock_singleton_failure(self):
-
+        """
+        Start a slow running thread and then try to acquire a shared
+        lock. Expect the acquire call to throw an exception.
+        """
         lock = LockSingleton('test')
```
Comment (posterzh) [inline hm_pyhelper/tests/test_lock_singleton.py:79]: @marvinmarnold I want to set the resource_count explicitly here. 
`lock = LockSingleton('test', resource_count=1)`

File: hm_pyhelper/lock_singleton.py (hunk 2 above)
Comment (posterzh) [reply hm_pyhelper/lock_singleton.py:11]: `InterprocessLock` sounds good. :+1:

**Comment by posterzh** [comment]
@marvinmarnold The available_resources of a `InterprocessLock` get reset on every restarting the system or docker container. It's tested in Ubuntu 20.04 desktop and diagnostics container in a Hotspot. I wrote about this in the Readme.md and comment section of `interprocess_lock.py`.

image

After restarting the `diagnostics` service.
image

File: hm_pyhelper/lock_singleton.py (hunk 1 above)
Comment (posterzh) [reply hm_pyhelper/lock_singleton.py]: Make sense. I refactored the this method to be easier to understand and added the comments how to use.

File: hm_pyhelper/lock_singleton.py (hunk 3 above)
Comment (posterzh) [reply hm_pyhelper/lock_singleton.py]: I added `timeout` parameter.

**Comment by vpetersson** [comment]
@posterzh What happens if the process crashes *after* the lock is set? Wouldn't we then get an infinite lock? I see there is a timeout for *acquiring* the lock, but not for resetting the lock.

**Review by rxt-rob — APPROVED** [review]

**Review by robputt — APPROVED** [review]
Nice :-).  I think all of Marvin's requested changes have been completed. 👍

**Review by marvinmarnold — APPROVED** [review]
I'll implement the changes in hm-diag now.

BTW @posterzh did we ever actually test that this does in fact lock across two or more different containers on balena?

**Pull request merged**